src/.clang-format

```
@@ -0,0 +1,35 @@
+BasedOnStyle: LLVM
+IndentWidth: 4
+TabWidth: 4
+UseTab: Never
+ColumnLimit: 120
+PenaltyBreakComment: 300
+PenaltyBreakFirstLessLess: 120
+PenaltyBreakString: 100
+PenaltyExcessCharacter: 100
+MaxEmptyLinesToKeep: 2
+BreakBeforeBraces: Attach
+AllowShortIfStatementsOnASingleLine: true
+AllowShortLoopsOnASingleLine: false
+AllowShortFunctionsOnASingleLine: false
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
+AlignTrailingComments: true
+PointerAlignment: Right
+KeepEmptyLinesAtTheStartOfBlocks: false
+SpaceBeforeParens: ControlStatements
+SpacesInParentheses: false
+SpacesInAngles: false
+SpacesInCStyleCastParentheses: false
+SpaceAfterCStyleCast: false
+SpacesInSquareBrackets: false
+ReflowComments: true
+CommentPragmas: '^\\s*\\*'
+InsertBraces: true
```

```suggestion
InsertBraces: false
```
I don't feel like we should force this, I see it is causing a lot of changes.

---

"optional syntax" in different places like un-braced if/while/do/for statements are an easy way to have bugs slip in by accident so it probably shouldn't be allowed. Much better for standard reliability development practices to enforce braces everywhere.

Typical arguments against it are:

- but we enforce it with formatting (weak argument in a non-space-sensitive language)
- but i don't want to (weak argument)
- but i want statements only on one line (weak argument, makes updating/refactoring difficult)

everybody, at some point, has done a change like this when in a hurry and broken something. The easy fix is to just _not allow it_ and enforce it with coding styles:

```c
if (thing)
  a += 1;

if (thing)
  printf("CHECK: %d\n", a);
  a += 1;
```

Or, you have a one line if statement and you want to paste something else in there for structure, so you have to re-add braces when it would cost nothing to just use them in the first place?

Also without braces, it makes quick "capture braces for cut/paste" moving/refactoring more difficult too.

---

On the good side, the existing code is so totally random with its braces that the programmer is probably ready for this sh*t.

---

```
if (thing)
  a += 1;
```
vs
```
if (thing) a+= 1;
```

Is actually the format I use fairly frequently, since I think when used appropriately it makes code a lot more readable. A case in module.c was:
```
        if (clusterNodeIsMyself(node)) *flags |= VALKEYMODULE_NODE_MYSELF;
        if (clusterNodeIsMaster(node)) *flags |= VALKEYMODULE_NODE_PRIMARY;
        if (clusterNodeIsSlave(node)) *flags |= VALKEYMODULE_NODE_REPLICA;
        if (clusterNodeTimedOut(node)) *flags |= VALKEYMODULE_NODE_PFAIL;
        if (clusterNodeIsFailing(node)) *flags |= VALKEYMODULE_NODE_FAIL;
        if (clusterNodeIsNoFailover(node)) *flags |= VALKEYMODULE_NODE_NOFAILOVER;
```

I could give it up though.

---

I would never complain about the last example in a code review.

If you asked me, I would vote against enforcing braces (or the equivalent for each language), but I know that nearly every coding standard in the world does insist on them.

If you are being forced to use braces, you can always write these examples as
~~~
if (x) { *flags |= y; }
~~~
Sadly, that too is disallowed by every set of coding standards I have ever seen.

The best advice I have ever seen came from a paper I read in the 90's.  It recognized that different people respond in different ways to the same thing, so we should store all programs in a canonical format, but display them and edit them in whatever format the user likes.  Sadly, even today our formatters are not good enough to support this.

---

I always use {} for statements that could be written on a single line - because if we ended up adding a new line there the change would be smaller and clearer. that said, I am totally fine with trading it in for consistency-via-clang-format.

---

> that said, I am totally fine with trading it in for consistency-via-clang-format.

I'm really not. I don't think this is worth enforcing. It's a huge chunk of the lines getting changed.

---

I'm skeptical to enabling this one.

> Warning
>
> Setting this option to true could lead to incorrect code formatting due to clang-format’s lack of complete semantic information. As such, extra care should be taken to review code changes made by this option.

Our code base often omits braces when if is on a single line, like `if (x) return y;`. (There's another config for that: AllowShortIfStatementsOnASingleLine). I'm fine with requiring braces if there's a line break after `if (x)` IMHO.

---

Warnings like that scare me.

If for some reason we choose to enable that option, then we really should take the recommended extra care.  But that would require that we get some kind of alert when it happens.  Does such an alert get generated by the formatter?

---

@madolson  "I don't feel like we should force this, I see it is causing a lot of changes." lets rip the band-aid once... 
besides, git now days knows to look behind marked commits and blame the correct person around the re-formatted code

So lets decide on what we like and less on what we have

---

My point is more that a lot of people prefer this. I don't like forcing style that is harder to read. I don't like forcing the braces.

---

If we want to introduce this incrementally, we should start with more liberal rules. Then we can make them more strict later on.

---

> If we want to introduce this incrementally, we should start with more liberal rules. Then we can make them more strict later on.

I like this.